compiler/rustc_hir_analysis/src/hir_ty_lowering/errors.rs

```diff
@@ -903,8 +903,8 @@ impl<'tcx> dyn HirTyLowerer<'tcx> + '_ {
                     format!("{}, {}>", &snippet[..snippet.len() - 1], types.join(", "))
                 } else {
                     // The user wrote `Iterator`, so we don't have a type we can suggest, but at
-                    // least we can clue them to the correct syntax `Iterator<Item = Type>`.
-                    format!("{}<{}>", snippet, types.join(", "))
+                    // least we can clue them to the correct syntax `Iterator::<Item = Type>`.
+                    format!("{}::<{}>", snippet, types.join(", "))
```

Comment:
We should only add `::` if the path (of the traitref) is located inside an expr or pat.
You should be able to obtain this information by checking if `tcx.parent_hir_node(bound.trait_ref.hir_id)` is `Node::Expr(_) | Node::Pat(_)` here: https://github.com/rust-lang/rust//blob/3ce7f9eb0c9ae2cdbc3433d738aa2842d56285e3/compiler/rustc_hir_analysis/src/hir_ty_lowering/errors.rs#L748-L782 and then threading this piece of information all the way to here.

Reply:
I don't think this works. Given [the most recent tests](<https://github.com/rust-lang/rust/pull/127094/commits/f8911e3f65c60e7ce83297b62e5f0ec18fdac421>) and print debugging like so:

```rust
let mut in_expr_or_pat = false;
if let ([], [bound]) = (&potential_assoc_types[..], &trait_bounds) {
    in_expr_or_pat = match tcx.parent_hir_node(bound.trait_ref.hir_ref_id) {
        Node::Expr(_) | Node::Pat(_) => true,
        _ => false,
    };
    println!(
        "(!!!!!!) parent node: {:#?}",
        tcx.parent_hir_node(bound.trait_ref.hir_ref_id)
    );
    ...
```

You can see that both instances in [my new dynless-turbofish test](<tests/ui/errors/dynless-turbofish-e0191-issue-91997.rs>) and [the already existing E0191 test](<tests/ui/error-codes/E0191.stderr>) have the same overarching `Ty` variant. I'll push this print debugging to this branch so you can see it, but it's of my current understanding that there should be some other way of checking this. Thanks.